#📌 Extracción

In [58]:
import pandas as pd

In [59]:
datos = pd.read_json('/content/TelecomX_Data.json')

In [60]:
datos.info()
datos.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

##Normalizando las columnas anidadas

In [61]:
#df_customer_normalize = pd.json_normalize(datos['customer'])
#df_phone_normalize = pd.json_normalize(datos['phone'])
#df_internet_normalize = pd.json_normalize(datos['internet'])
#df_account_normalize = pd.json_normalize(datos['account'])

In [62]:
datos1 = [pd.json_normalize(datos[col]) for col in ['customer', 'phone', 'internet', 'account']]

##Creando el DataFrame con todas las columnas

In [63]:
#df_normalized = pd.concat([datos[['customerID']],datos[['Churn']],df_customer_normalize, df_phone_normalize, df_internet_normalize, df_account_normalize], axis=1)
#df_normalized.head()

In [64]:
df_final = pd.concat([datos[['customerID']],datos[['Churn']],*datos1], axis=1)
df_final.head()


customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

In [65]:
list(df_final.columns)


['customerID',
 'Churn',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'Charges.Monthly',
 'Charges.Total']

##Diccionario de datos

*    customerID: número de identificación único de cada cliente
*    Churn: si el cliente dejó o no la empresa
*    gender: género (masculino y femenino)
*    SeniorCitizen: información sobre si un cliente tiene o no una edad    igual o mayor a 65 años
*    Partner: si el cliente tiene o no una pareja
*    Dependents: si el cliente tiene o no dependientes
*    tenure: meses de contrato del cliente
*    PhoneService: suscripción al servicio telefónico
*    MultipleLines: suscripción a más de una línea telefónica
*   InternetService: suscripción a un proveedor de internet
*    OnlineSecurity: suscripción adicional de seguridad en línea
*    OnlineBackup: suscripción adicional de respaldo en línea
*    DeviceProtection: suscripción adicional de protección del dispositivo
*    TechSupport: suscripción adicional de soporte técnico, menor tiempo de espera
*    StreamingTV: suscripción de televisión por cable
*    StreamingMovies: suscripción de streaming de películas
*    Contract: tipo de contrato
*    PaperlessBilling: si el cliente prefiere recibir la factura en línea
*    PaymentMethod: forma de pago
*    Charges.Monthly: total de todos los servicios del cliente por mes
*    Charges.Total: total gastado por el cliente



#🔧 Transformación

###Columnas en letra minuscula

In [66]:
df_final.columns = df_final.columns.str.lower()

###Columnas sin espacios

In [67]:
df_final.columns = df_final.columns.str.strip()

In [68]:
df_final.columns

Index(['customerid', 'churn', 'gender', 'seniorcitizen', 'partner',
       'dependents', 'tenure', 'phoneservice', 'multiplelines',
       'internetservice', 'onlinesecurity', 'onlinebackup', 'deviceprotection',
       'techsupport', 'streamingtv', 'streamingmovies', 'contract',
       'paperlessbilling', 'paymentmethod', 'charges.monthly',
       'charges.total'],
      dtype='object')

###Renombrando las columnas

In [69]:
df_final = df_final.rename(columns={'seniorcitizen': 'senior_citizen', 'phoneservice': 'phone_service',
                                    'multiplelines': 'multiple_lines', 'internetservice': 'internet_service',
                                    'onlinesecurity': 'online_security', 'onlinebackup': 'online_backup',
                                    'deviceprotection': 'device_protection', 'techsupport': 'tech_support',
                                    'streamingtv': 'streaming_tv', 'streamingmovies': 'streaming_movies',
                                    'paperlessbilling': 'paperless_billing', 'paymentmethod': 'payment_method',
                                    'charges.monthly': 'monthly_charges', 'charges.total': 'total_charges'})



In [70]:
df_final.columns

Index(['customerid', 'churn', 'gender', 'senior_citizen', 'partner',
       'dependents', 'tenure', 'phone_service', 'multiple_lines',
       'internet_service', 'online_security', 'online_backup',
       'device_protection', 'tech_support', 'streaming_tv', 'streaming_movies',
       'contract', 'paperless_billing', 'payment_method', 'monthly_charges',
       'total_charges'],
      dtype='object')

In [71]:
df_final.info()
df_final.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customerid         7267 non-null   object 
 1   churn              7267 non-null   object 
 2   gender             7267 non-null   object 
 3   senior_citizen     7267 non-null   int64  
 4   partner            7267 non-null   object 
 5   dependents         7267 non-null   object 
 6   tenure             7267 non-null   int64  
 7   phone_service      7267 non-null   object 
 8   multiple_lines     7267 non-null   object 
 9   internet_service   7267 non-null   object 
 10  online_security    7267 non-null   object 
 11  online_backup      7267 non-null   object 
 12  device_protection  7267 non-null   object 
 13  tech_support       7267 non-null   object 
 14  streaming_tv       7267 non-null   object 
 15  streaming_movies   7267 non-null   object 
 16  contract           7267 

customerid churn  gender  senior_citizen partner dependents  tenure  \
0  0002-ORFBO    No  Female               0     Yes        Yes       9   
1  0003-MKNFE    No    Male               0      No         No       9   
2  0004-TLHLJ   Yes    Male               0      No         No       4   
3  0011-IGKFF   Yes    Male               1     Yes         No      13   
4  0013-EXCHZ   Yes  Female               1     Yes         No       3   

  phone_service multiple_lines internet_service  ... online_backup  \
0           Yes             No              DSL  ...           Yes   
1           Yes            Yes              DSL  ...            No   
2           Yes             No      Fiber optic  ...            No   
3           Yes             No      Fiber optic  ...           Yes   
4           Yes             No      Fiber optic  ...            No   

  device_protection tech_support streaming_tv streaming_movies  \
0                No          Yes          Yes               No   
1                No           No           No              Yes   
2               Yes           No           No               No   
3               Yes           No          Yes              Yes   
4                No          Yes          Yes               No   

         contract paperless_billing    payment_method monthly_charges  \
0        One year               Yes      Mailed check            65.6   
1  Month-to-month                No      Mailed check            59.9   
2  Month-to-month               Yes  Electronic check            73.9   
3  Month-to-month               Yes  Electronic check            98.0   
4  Month-to-month               Yes      Mailed check            83.9   

   total_charges  
0          593.3  
1          542.4  
2         280.85  
3        1237.85  
4          267.4  

[5 rows x 21 columns]

###Valores unicos de las columnas

In [72]:
columnas = list(df_final.columns)

valores_unicos = {}

for columna in columnas:
    print(f'{columna}: {df_final[columna].unique()}')


customerid: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
churn: ['No' 'Yes' '']
gender: ['Female' 'Male']
senior_citizen: [0 1]
partner: ['Yes' 'No']
dependents: ['Yes' 'No']
tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 11 37 49 66
 67 20 43 59 12 27  2 25 29 14 35 64 39 40  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 51 31 36 17 18 38 42
  0]
phone_service: ['Yes' 'No']
multiple_lines: ['No' 'Yes' 'No phone service']
internet_service: ['DSL' 'Fiber optic' 'No']
online_security: ['No' 'Yes' 'No internet service']
online_backup: ['Yes' 'No' 'No internet service']
device_protection: ['No' 'Yes' 'No internet service']
tech_support: ['Yes' 'No' 'No internet service']
streaming_tv: ['Yes' 'No' 'No internet service']
streaming_movies: ['No' 'Yes' 'No internet service']
contract: ['One year' 'Month-to-month' 'Two year']
paperless_billing: ['Yes' 'No']
payment_method: ['Mailed check' 'Electroni

##Determinando Valores vacios en la columna churn

In [73]:
df_final['churn'].value_counts()

churn
No     5174
Yes    1869
        224
Name: count, dtype: int64

In [74]:
vacios = (df_final['churn'] == '').sum()
vacios


np.int64(224)

###Porcentaje de los valores vacios

In [75]:
total_filas_churn = len(df_final)
porcentaje = (df_final['churn'].value_counts() / total_filas_churn) * 100
print(f'{porcentaje.iloc[0]:.2f}%')
print(f'{porcentaje.iloc[1]:.2f}%')
print(f'{porcentaje.iloc[2]:.2f}% Porcentaje de valores vacios ')

71.20%
25.72%
3.08% Porcentaje de valores vacios 


##Vemos que la columna 'churn' contiene 224 datos vacios
**Por que eliminar**:
* Esto representa el 3.08% del total de datos de la columna.
* Es un porcentaje bajo < 5%, no afectara significativamente el analisis.
* Evitamos sesgos en modelos predictivos.

###Elimina los valores vacios '' en la columna churn

In [78]:
df_clean = df_final[df_final['churn'] != ''].copy()
df_clean['churn'].value_counts()

churn
No     5174
Yes    1869
Name: count, dtype: int64

In [80]:
columnas = list(df_clean.columns)

valores_unicos = {}

for columna in columnas:
    print(f'{columna}: {df_clean[columna].unique()}')

customerid: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
churn: ['No' 'Yes']
gender: ['Female' 'Male']
senior_citizen: [0 1]
partner: ['Yes' 'No']
dependents: ['Yes' 'No']
tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 37 49 66 67
 20 43 59 12 27  2 25 29 14 35 64 39 40 11  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 31 36 17 18 51 38 42
  0]
phone_service: ['Yes' 'No']
multiple_lines: ['No' 'Yes' 'No phone service']
internet_service: ['DSL' 'Fiber optic' 'No']
online_security: ['No' 'Yes' 'No internet service']
online_backup: ['Yes' 'No' 'No internet service']
device_protection: ['No' 'Yes' 'No internet service']
tech_support: ['Yes' 'No' 'No internet service']
streaming_tv: ['Yes' 'No' 'No internet service']
streaming_movies: ['No' 'Yes' 'No internet service']
contract: ['One year' 'Month-to-month' 'Two year']
paperless_billing: ['Yes' 'No']
payment_method: ['Mailed check' 'Electronic c

###Convertir columnas especificas de (Yes) = 1 y (No) = 0

In [85]:
df_clean[['churn', 'partner', 'dependents', 'phone_service', 'paperless_billing']] = df_clean[['churn', 'partner', 'dependents', 'phone_service', 'paperless_billing']].replace({'Yes': 1, 'No': 0}).astype(int)

df_clean = df_clean.reset_index(drop=True)
df_clean.info()
df_clean.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customerid         7043 non-null   object 
 1   churn              7043 non-null   int64  
 2   gender             7043 non-null   object 
 3   senior_citizen     7043 non-null   int64  
 4   partner            7043 non-null   int64  
 5   dependents         7043 non-null   int64  
 6   tenure             7043 non-null   int64  
 7   phone_service      7043 non-null   int64  
 8   multiple_lines     7043 non-null   object 
 9   internet_service   7043 non-null   object 
 10  online_security    7043 non-null   object 
 11  online_backup      7043 non-null   object 
 12  device_protection  7043 non-null   object 
 13  tech_support       7043 non-null   object 
 14  streaming_tv       7043 non-null   object 
 15  streaming_movies   7043 non-null   object 
 16  contract           7043 

customerid  churn  gender  senior_citizen  partner  dependents  tenure  \
0  0002-ORFBO      0  Female               0        1           1       9   
1  0003-MKNFE      0    Male               0        0           0       9   
2  0004-TLHLJ      1    Male               0        0           0       4   
3  0011-IGKFF      1    Male               1        1           0      13   
4  0013-EXCHZ      1  Female               1        1           0       3   

   phone_service multiple_lines internet_service  ... online_backup  \
0              1             No              DSL  ...           Yes   
1              1            Yes              DSL  ...            No   
2              1             No      Fiber optic  ...            No   
3              1             No      Fiber optic  ...           Yes   
4              1             No      Fiber optic  ...            No   

  device_protection tech_support streaming_tv streaming_movies  \
0                No          Yes          Yes               No   
1                No           No           No              Yes   
2               Yes           No           No               No   
3               Yes           No          Yes              Yes   
4                No          Yes          Yes               No   

         contract paperless_billing    payment_method monthly_charges  \
0        One year                 1      Mailed check            65.6   
1  Month-to-month                 0      Mailed check            59.9   
2  Month-to-month                 1  Electronic check            73.9   
3  Month-to-month                 1  Electronic check            98.0   
4  Month-to-month                 1      Mailed check            83.9   

   total_charges  
0          593.3  
1          542.4  
2         280.85  
3        1237.85  
4          267.4  

[5 rows x 21 columns]

#📊 Carga y análisis

#📄Informe final